### PR TITLE
[risk=no][RW-4140] Puppeteer automation

### PR DESCRIPTION
**Pull request by aweng98**
[Proposal doc](https://docs.google.com/document/d/1ERllaxKD8D1SE5bNMfMQ2-_pyfG8_IpvRQVwNSMPYxc/edit?usp=sharing)

A basic Puppeteer test automation framework in JS for the AoU team UI and integration testing.
[RW-4140](https://precisionmedicineinitiative.atlassian.net/browse/RW-4140)

### CODE REVIEW

File: automation/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
+# Run all UI tests
+- cd automation/
+- npm test
```
Comment (blrubenstein) [inline automation/README.md]: Can we run through yarn? Just to stay consistent.

File: automation/.circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      - run: yarn install
+
+      - save_cache:
+          paths:
+            - node_modules
+          key: v1-dependencies-{{ checksum "package.json" }}
+     
+      # run tests 
+      - run: yarn run lint
+      - run: yarn run test
```
Comment (blrubenstein) [inline automation/.circleci/config.yml]: Will this trigger off our builds/workflows in `.circleci`?

File: automation/src/configs/app.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
+  "pages": {
+    "login": "/login",
+    "workspaces": "/workspaces",
+    "profile": "/profile",
+    "library": "/library",
+    "admin": "/admin/user"
+  },
+  "test": {
```
Comment (blrubenstein) [inline automation/src/configs/app.json]: High level question that may be worth considering: can we have it spin up a local server and run tests against that? I could see that being highly valuable so we could run tests on every PR.

File: automation/README.md (hunk 1 above)
Comment (aweng98) [reply automation/README.md]: File is not updated. yarn test works and it’s what I use to run test.

File: automation/src/configs/app.json (hunk 3 above)
Comment (aweng98) [reply automation/src/configs/app.json]: Yes that’s the plan. switching env will be done via a node property which doesn’t exist atm. I will add it today

File: automation/.eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+module.exports = {
```
Comment (aweng98) [inline automation/.eslintrc.js]: eslint not used anymore. I forgot to remove this. See tslint.

File: automation/.circleci/config.yml (hunk 2 above)
Comment (aweng98) [reply automation/.circleci/config.yml]: I haven’t tested this out. It’s roughly a template for now

File: automation/.env_template [hunk 5]
```diff
@@ -0,0 +1,3 @@
+USER_NAME=??
```
Comment (aweng98) [inline automation/.env_template]: Copy and rename new file as ‘.env’ and update with testuser login credential

File: automation/README.md (hunk 1 above)
Comment (blrubenstein) [reply automation/README.md]: 👍 sounds good.

File: automation/jest.config.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  "verbose": true,
+  "globals": {
+    "ts-jest": {
+      "tsConfig": "tsconfig.jest.json"
+    }
+  },
+  "testPathIgnorePatterns": [
+    "/node_modules/"
+  ],
+  // "testRegex": ["\\.(test|spec)\\.tsx?$"],
```
Comment (jaycarlton) [inline automation/jest.config.js]: please remove the commented-out code.

File: automation/src/data/MockData.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import faker from 'faker';
+
+/**
+ * Get fake user information.
+ */
+const user = {
```
Comment (jaycarlton) [inline automation/src/data/MockData.ts]: nit: I like to name my mocks like `mockUser` so that when tests start to get complex, you can tell what's real and what's not.

File: automation/src/data/MockData.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  firstname: faker.name.firstName(),
+  lastname: faker.name.lastName(),
+  job: faker.name.jobTitle(),
+  company: faker.company(),
+  words: faker.random.words(5),
```
Comment (jaycarlton) [inline automation/src/data/MockData.ts]: Let's keep the input the same every time, so `faker.lorem.words(5)`. I generally don't want any randomness on the input side in a test.

File: automation/src/configs/app.json (hunk 3 above)
Comment (aweng98) [reply automation/src/configs/app.json]: it's done

File: automation/.eslintrc_backup.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
```
[the remark this answers is not in the document]
Comment (aweng98) [reply automation/.eslintrc_backup.js]: removed.

File: automation/.env_template [hunk 10]
```diff
@@ -0,0 +1,3 @@
...
+PASSWORD=??
+INVITATION_KEY=??
```
[the remark this answers is not in the document]
Comment (aweng98) [reply automation/.env_template]: done!

File: automation/src/fixtures/chrome.ts [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    '--disable-accelerated-2d-canvas',
+    '--disable-gpu',
+    '--enable-features=NetworkService' // https://github.com/GoogleChrome/puppeteer/issues/1664
+  ]
+};
+
+class Chrome {
+
+  async launch(opts = undefined) {
+    const launchOptions = (opts === undefined) ? defaultOpts : opts;
```
[the remark this answers is not in the document]
Comment (aweng98) [reply automation/src/fixtures/chrome.ts]: 👍

File: automation/jest.config.js (hunk 6 above)
Comment (aweng98) [reply automation/jest.config.js]: done!

File: automation/src/data/MockData.ts (hunk 7 above)
Comment (aweng98) [reply automation/src/data/MockData.ts]: renamed to `fakeUser`

File: automation/src/pages/mixin/NavigationMenu.ts [hunk 12]
```diff
@@ -0,0 +1,156 @@
+import {ElementHandle, Page} from 'puppeteer-core';
+
+/**
+ * xpath selectors
+ */
+const selectors = {
+  root: '//',
+  hamburgerLink: '//div[contains(@style,\'transform: rotate(90deg)\') or contains(@style,\'transform: rotate(0deg)\')]/*[@shape=\'bars\']',
```
Comment (calbach) [inline automation/src/pages/mixin/NavigationMenu.ts]: These don't look stable to me. We should have further discussion about how we want to select elements

File: automation/src/pages/HomePage.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    return this;
+  }
+
+  get seeAllWorkspacesLink(): Promise<ElementHandle> {
+    return this.puppeteerPage.waitForXPath(selectors.seeAllWorkspacesLink, {visible: true})
+  }
+
+  public async toAllWorkspaces() {
+    const link = await this.seeAllWorkspacesLink;
+    await Promise.all([link.click(), this.puppeteerPage.waitForNavigation({waitUntil: 'networkidle0'})]);
```
Comment (panentheos) [inline automation/src/pages/HomePage.ts]: Is it necessary to explicitly wait for navigation? If we just wait for the next interaction directly, it should cross page boundaries without having to do that manually.

File: automation/src/tests/navigation/navigation-menu.spec.ts [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+  test('User open all-workspaces page', async () => {
+    const homePage = new HomePage(page);
+    await homePage.waitForReady();
+
+    await homePage.navigation.toAllWorkspaces();
+
+    const workspaces = new WorkspacesPage(page);
+    await workspaces.waitForReady();
+    expect(await page.url()).toContain('workspaces');
```
Comment (panentheos) [inline automation/src/tests/navigation/navigation-menu.spec.ts]: Small, single-interaction tests like this tend not to provide much value by themselves. Are we expecting that the 'See all workspaces' link is at such high risk of breaking that it needs an isolated test case?

A better use of this kind of automation is to string together longer flows that mimic how a user would actually interact with the site. Those kinds of tests can uncover bugs and edge cases that live within the complexity of the system as a whole, rather than simple coding mistakes that are easier to catch by other means.

File: automation/src/pages/mixin/NavigationMenu.ts (hunk 12 above)
Comment (panentheos) [reply automation/src/pages/mixin/NavigationMenu.ts]: To add to this: If it's not straightforward to write a text-based selector for an element, it means that element is probably inaccessible to someone using e.g. a screen reader. This is a good signal that we should probably add an `aria-label` or some other text labeling here.

File: automation/src/services/element-handler.ts [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+/**
+ * <pre>
+ *  Check if the element is visible
+ * </pre>
+ * @param {Page} page
+ * @param {ElementHandle} element
+ */
+export async function visible(page: Page, element: ElementHandle) {
```
Comment (gjuggler) [inline automation/src/services/element-handler.ts]: naming nitpick: isVisible? It just feels more clearly like an assertion type method that way.

**Comment by calbach** [comment]
I'm not sure 7 reviewers on a large PR is a good idea. I'd suggesting popping up a level and establishing the basic outline of the testing framework for broader high-level review (perhaps via the proposal doc) before we do a detailed code review with a smaller set of reviewers. In particular addressing auth and DOM selection patterns.

File: puppeteer-test/src/tests/navigation/navigationmenu.spec.ts [hunk 16]
```diff
@@ -0,0 +1,68 @@
+import GoogleLoginPage from '../../pages/google-login';
```
Comment (gjuggler) [inline puppeteer-test/src/tests/navigation/navigationmenu.spec.ts]: It may be more maintainable to use absolute paths rather than relative paths, if possible. This is what our UI codebase uses in most cases... if it's not possible here for some reason, I'd love to know why (and quickly see if we can fix it).

File: puppeteer-test/src/tests/navigation/navigationmenu.spec.ts [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+  let browser;
+  let incognitoContext;
+  let page;
+
+  beforeAll(async () =>  {
+    browser = await PuppeteerLaunch();
+  });
+
+  beforeEach(async () => {
+    incognitoContext = await browser.createIncognitoBrowserContext();
```
Comment (gjuggler) [inline puppeteer-test/src/tests/navigation/navigationmenu.spec.ts]: Is the Incognito context piece important here? It's not used for the homepage spec, so I'm wondering about the inconsistency. If I'm a RW dev planning to write a new spec for my new feature, which pattern should I prefer? Documenting the rationale behind this choice somewhere would be helpful – this could equally end up in a "base" test case wrapper file, in the README, or as a comment in one of these files.

File: puppeteer-test/src/tests/user/login.spec.ts [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    const googleButton = await loginPage.loginButton;
+    await googleButton.click();
+    await naviPromise;
+
+    await loginPage.enterEmail(configs.userEmail);
+    await loginPage.enterPassword(INCORRECT_PASSWORD);
+    const button = await page.waitForXPath(selectors.passwordNextButton);
+    await button.click();
+
+    const puppePage = new BasePage(page);
```
Comment (gjuggler) [inline puppeteer-test/src/tests/user/login.spec.ts]: spelling nitpick: puppeteerPage

File: puppeteer-test/src/tests/workspace/workspace-create.spec.ts [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+  afterEach(async () => {
+    await incognitoContext.close();
+  });
+
+  afterAll(async () => {
+    await browser.close();
+  });
+
+  test('Create a new workspace from the Home page', async () => {
+    const loginPage = new GoogleLoginPage(page);
```
Comment (gjuggler) [inline puppeteer-test/src/tests/workspace/workspace-create.spec.ts]: I think it might be meaningful to create a helper function that handles Google login automatically, and simply allows a test case to say "Get me to page X, using whatever auth flow is necessary". This may help reduce boilerplate and allow us flexibility to swap-out the user/email password entry with an alternate "backdoor" accessToken auth mechanism in the future.

File: puppeteer-test/src/tests/workspace/workspace-permission.spec.ts [hunk 20]
```diff
@@ -0,0 +1,39 @@
+import PuppeteerLaunch from '../../services/puppeteer-launch';
```
Comment (gjuggler) [inline puppeteer-test/src/tests/workspace/workspace-permission.spec.ts]: What are workspace-permission tests intended to be? It may not be worth keeping this around, if it's easy enough for folks to create the new file as needed.

File: puppeteer-test/tslint.json [hunk 21]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (gjuggler) [inline puppeteer-test/tslint.json]: [Opt / no action required] Is there some way we can automatically inherit / share our tslint with the /ui/ directory? It feels like linting rules should be treated as a universal style enforcer for our wider codebase.

**Comment by gjuggler** [comment]
Oh – one add'l comment / request: please update the PR description with a ticket and risk score, and remove the [work in progress] prefix :)

File: puppeteer-test/src/tests/navigation/navigationmenu.spec.ts (hunk 17 above)
Comment (aweng98) [reply puppeteer-test/src/tests/navigation/navigationmenu.spec.ts]: I recommend all tests run in Incognito mode to prevent tests sharing any cookies or cache with the other tests, especially true if running in parallel. The Homepage spec is a very simple version of example test created at the beginning. I will remove Homepage spec very soon in near future.

File: puppeteer-test/src/config/config.js [hunk 22]
```diff
@@ -0,0 +1,61 @@
+require('dotenv').config(); // loads variables from .env
+const _ = require('lodash');
+
+const env = process.env.NODE_ENV || 'test'; // default env is test
```
[the remark this answers is not in the document]
Comment (aweng98) [reply puppeteer-test/src/config/config.js]: `WORKBENCH_ENV` is better. 👍 Changed.

File: puppeteer-test/jest.config.js [hunk 23]
```diff
@@ -0,0 +1,24 @@
+module.exports = {
```
[the remark this answers is not in the document]
Comment (aweng98) [reply puppeteer-test/jest.config.js]: Done. Deleted jest.config.js

File: puppeteer-test/README.md [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+
+### We use `yarn` instead `npm`
+#### Install all libs
+- `yarn`
+
+#### List all `yarn` tasks from `package.json`
+- `yarn run`
+
+#### Run all UI tests found in `/tests` folder.
+- Run command: `cd automation/`
```
[the remark this answers is not in the document]
Comment (aweng98) [reply puppeteer-test/README.md]: I try to keep dir layers as small and simple as possible, quickier to find. How about `automation-puppeteer`?

File: puppeteer-test/src/tests/workspace/workspace-permission.spec.ts (hunk 20 above)
Comment (aweng98) [reply puppeteer-test/src/tests/workspace/workspace-permission.spec.ts]: Deleted

File: puppeteer-test/tslint.json (hunk 21 above)
Comment (aweng98) [reply puppeteer-test/tslint.json]: for now `tsconfig.json` extends `../ui/tsconfig.json`. I don't know how to extends tslint but I'll look into it.

**Comment by aweng98** [comment]
I like "e2e", Done.
Good suggestion on marking comments as "Done" in future. Thank you.
